pytest_html/plugin.py

```diff
@@ -127,18 +136,44 @@ def __lt__(self, other):
                      'XPassed', 'Skipped', 'Passed')
             return order.index(self.outcome) < order.index(other.outcome)
 
-        def append_extra_html(self, extra, additional_html, links_html):
+        def append_extra_html(self, extra, extra_index, test_index):
             href = None
             if extra.get('format') == extras.FORMAT_IMAGE:
-                href = '#'
-                image = 'data:image/png;base64,{0}'.format(
-                        extra.get('content'))
-                additional_html.append(html.div(
-                    html.a(html.img(src=image), href="#"),
+                if self.self_contained:
+                    image_src = 'data:image/png;base64,{0}'.format(
+                            extra.get('content'))
+                    href = '#'
+                else:
+                    hash_key = ''.join([self.test_id, str(extra_index),
+                                       str(test_index)]).encode('utf-8')
+                    hash_generator = hashlib.md5()
+                    hash_generator.update(hash_key)
+                    image_file_name = '{0}.png'.format(hash_generator
+                                                       .hexdigest())
+                    image_path = os.path.join(os.path.dirname(self.logfile),
+                                              'assets', image_file_name)
+                    if not os.path.exists(os.path.dirname(image_path)):
+                        os.makedirs(os.path.dirname(image_path))
+
+                    href = os.path.relpath(image_path,
+                                           os.path.dirname(self.logfile))
+                    image_src = href
+
+                    with open(image_path, 'w') as f:
+                        content = extra.get('content')
+                        if PY3:
+                            content = bytearray()
+                            content.extend(extra.get('content')
+                                           .encode('utf-8'))
```

Comment:
It looks like `bytearray` can be [initialized](https://docs.python.org/3/library/functions.html#bytearray) with `source` and `encoding`.